crates/bindings/src/lib.rs

```diff
@@ -45,13 +42,610 @@ pub use timestamp::Timestamp;
 
 pub type ReducerResult = core::result::Result<(), Box<str>>;
 
-/// A context that any reducer is provided with.
+pub use spacetimedb_bindings_macro::duration;
+
+/// Generates code for registering a row-level security rule.
+///
+/// This attribute must be applied to a `const` binding of type [`Filter`].
+/// It will be interpreted as a filter on the table to which it applies, for all client queries.
+/// If a module contains multiple `client_visibility_filter`s for the same table,
+/// they will be unioned together as if by SQL `OR`,
+/// so that any row permitted by at least one filter is visible.
+///
+/// The `const` binding's identifier must be unique within the module.
+///
+/// The query follows the same syntax as a subscription query.
+///
+/// ## Example:
+///
+/// ```rust,ignore
+/// /// Players can only see what's in their chunk
+/// #[spacetimedb::client_visibility_filter]
+/// const PLAYERS_SEE_ENTITIES_IN_SAME_CHUNK: Filter = Filter::Sql("
+///     SELECT * FROM LocationState WHERE chunk_index IN (
+///         SELECT chunk_index FROM LocationState WHERE entity_id IN (
+///             SELECT entity_id FROM UserState WHERE identity = @sender
+///         )
+///     )
+/// ");
+/// ```
+///
+/// Queries are not checked for syntactic or semantic validity
+/// until they are processed by the SpacetimeDB host.
+/// This means that errors in queries, such as syntax errors, type errors or unknown tables,
+/// will be reported during `spacetime publish`, not at compile time.
+#[doc(inline, hidden)] // TODO: RLS filters are currently unimplemented, and are not enforced.
+pub use spacetimedb_bindings_macro::client_visibility_filter;
+
+/// Declares a table with a particular row type.
+///
+/// This attribute is applied to a struct type with named fields.
+/// This derives [`Serialize`], [`Deserialize`], [`SpacetimeType`], and [`Debug`] for the annotated struct.
+///
+/// Elements of the struct type are NOT automatically inserted into any global table.
+/// They are regular structs, with no special behavior.
+/// In particular, modifying them does not automatically modify the database!
+///
+/// Instead, a type implementing [`Table<Row = Self>`] is generated. This can be looked up in a [`ReducerContext`]
+/// using `ctx.db.{table_name}()`. This type represents a handle to a database table, and can be used to
+/// iterate and modify the table's elements. It is a view of the entire table -- the entire set of rows at the time of the reducer call.
+///
+/// # Example
+///
+/// ```ignore
+/// use spacetimedb::{table, ReducerContext};
+///
+/// #[table(name = users, public,
```

Comment:
```suggestion
/// #[table(name = user, public,
```

Table names should be singular.

Reply:
The rest of this doc comment will need to be updated too, but I'm too lazy to add suggestions to all of the uses.